godot-macros/src/util/list_parser.rs

```diff
@@ -0,0 +1,206 @@
+                let list_stream = list.iter().cloned().collect::<TokenStream>();
+                return bail!(list_stream, "expected expression, found `,`");
+            }
+
+            let end = if is_last { list.len() } else { list.len() - 1 };
+
+            lists.push(KvValue::new((&list[..end]).into()));
+        }
+
+        Ok(Self {
+            lists: lists.into(),
+            span_close: group.span_close(),
+        })
+    }
+
+    /// Get the next element from the list, starting at the front.
+    fn pop_next(&mut self) -> Option<KvValue> {
+        self.lists.pop_front()
+    }
+
+    fn peek(&self) -> Option<&KvValue> {
+        self.lists.front()
+    }
+
+    /// Take the next element of the list, ensuring it is an expression.
+    pub fn next_expr(&mut self) -> ParseResult<TokenStream> {
+        let Some(kv) = self.pop_next() else {
+            return bail!(self.span_close, "expected expression");
+        };
+
+        kv.expr()
+    }
+
+    /// Take the next element of the list, if it is an identifier.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left.
+    pub fn next_ident(&mut self) -> ParseResult<Option<Ident>> {
+        let Some(kv) = self.pop_next() else { return Ok(None) };
+
+        Ok(Some(kv.ident()?))
+    }
+
+    /// Take the next element of the list, if it is an identifier.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left or the next element isn't an identifier.
+    pub fn try_next_ident(&mut self) -> ParseResult<Option<Ident>> {
+        let Some(kv) = self.peek() else { return Ok(None) };
+
+        let id = kv.as_ident()?;
+
+        _ = self.pop_next();
+
+        Ok(Some(id))
+    }
+
+    /// Take the next element of the list, ensuring it is one of the given identifiers.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left.
+    pub fn next_any_ident(&mut self, ids: &[&str]) -> ParseResult<Option<Ident>> {
+        let Some(next_id) = self.try_next_ident()? else {
+            return Ok(None)
+        };
+
+        for id in ids {
+            if ident(id) == next_id {
+                return Ok(Some(next_id));
+            }
+        }
+
+        let allowed_values = ids.join(",");
+        bail!(next_id, "expected one of: \"{allowed_values}\"")
+    }
+
+    /// Take the next element of the list, if it is a key-value pair of the form `key = expression`.
+    pub(crate) fn try_next_key_value(&mut self) -> Option<(Ident, KvValue)> {
+        let Some(kv) = self.peek() else { return None };
+
+        if let Ok((key, value)) = kv.as_key_value() {
+            _ = self.pop_next();
+
+            Some((key, value))
+        } else {
+            None
+        }
+    }
+
+    /// Take the next element of the list, ensuring it is either a single identifier or a key-value pair of
+    /// the form `key = expression`.
+    pub(crate) fn next_key_optional_value(
+        &mut self,
+    ) -> ParseResult<Option<(Ident, Option<KvValue>)>> {
```

Comment:
This signature got quite complex with triply nested result/option/result.

Do we really hit all those code paths?

Reply:
We should:
- `Err(_)` should be hit by `#[export(enum = ("hello"))]`
- `Ok(None)` should be hit by `#[export(enum = ())]` (or any other case where we are done with parsing all key-value pairs)
- `Ok(Some((key, None)))` should be hit by `#[export(enum = (A))]`
- `Ok(Some((key, Some(value))))` should be hit by `#[export(enum = (A = 5))]`